### ISSUE TITLE
LDS cannot find new cluster because it's still warming up

### ISSUE THREAD

**Issue by winguse** [envoyproxy/envoy#5933]
We are using an ADS service to deliver LDS/CDS to envoy. We found in some cases, it will fail to add new cluster because of LDS complaining about the newly added cluster is unknown.

Here is the sequence diagram of added a new cluster:

![image](https://user-images.githubusercontent.com/1443504/52702879-138ce700-2fb8-11e9-87ee-56c0dcdc7bb8.png)

So:

1. Is that possible to make Envoy ack CDS after warming is done?
2. Or how do we decide what is the correct time for delivering LDS update to Envoy?

Thanks!

Envoy version: 1.9.0

Envoy Log:

```
[2019-02-13 09:04:40.313][000007][info][upstream] [source/common/upstream/cluster_manager_impl.cc:495] add/update cluster zipkin during init
[2019-02-13 09:04:40.315][000007][info][upstream] [source/common/upstream/cluster_manager_impl.cc:495] add/update cluster local_cluster during init
[2019-02-13 09:04:40.315][000007][info][upstream] [source/common/upstream/cluster_manager_impl.cc:495] add/update cluster kinesis_cluster during init
[2019-02-13 09:04:40.341][000007][info][upstream] [source/common/upstream/cluster_manager_impl.cc:136] cm init: all clusters initialized
[2019-02-13 09:04:40.341][000007][info][main] [source/server/server.cc:435] all clusters initialized. initializing init manager
[2019-02-13 09:04:40.355][000007][info][upstream] [source/server/lds_api.cc:80] lds: add/update listener 'ranking_ingress'
[2019-02-13 09:04:40.355][000007][info][upstream] [source/server/lds_api.cc:80] lds: add/update listener 'ranking_egress'
[2019-02-13 09:04:40.356][000007][info][upstream] [source/server/lds_api.cc:80] lds: add/update listener 'kinesis_egress'
[2019-02-13 09:04:40.356][000007][info][config] [source/server/listener_manager_impl.cc:961] all dependencies initialized. starting workers
[2019-02-13 09:04:46.297][000007][info][upstream] [source/common/upstream/cluster_manager_impl.cc:501] add/update cluster datascience starting warming
[2019-02-13 09:04:46.304][000007][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_mux_subscription_lib/common/config/grpc_mux_subscription_impl.h:70] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Error adding/updating listener ranking_egress: route: unknown cluster 'datascience'
```

**Comment by ramaraochavali** [comment]
Related issue https://github.com/envoyproxy/envoy/issues/5174 - if it is implemented, I think it should address this case as well?

**Comment by htuch** [comment]
Yes, this is a dupe of #5174, so will close.

**Comment by winguse** [comment]
@htuch I think this is not fixed. I tested with envoy `1.13.1`, there is possibility it will happen:

```
envoy_1  | [2020-03-17 07:23:12.245][1][info][upstream] [source/common/upstream/cds_api_impl.cc:74] cds: add 7 cluster(s), remove 1 cluster(s)
envoy_1  | [2020-03-17 07:23:12.246][1][info][upstream] [source/common/upstream/cds_api_impl.cc:90] cds: add/update cluster 'kinesis_cluster'
envoy_1  | [2020-03-17 07:23:12.248][1][info][upstream] [source/common/upstream/cds_api_impl.cc:90] cds: add/update cluster 'mercury'
envoy_1  | [2020-03-17 07:23:12.249][1][info][upstream] [source/common/upstream/cds_api_impl.cc:90] cds: add/update cluster 'chameleon'
envoy_1  | [2020-03-17 07:23:12.251][1][info][upstream] [source/common/upstream/cds_api_impl.cc:90] cds: add/update cluster 'dmp'
envoy_1  | [2020-03-17 07:23:12.262][1][info][upstream] [source/server/lds_api.cc:55] lds: remove listener 'jenkins_test_egress'
envoy_1  | [2020-03-17 07:23:12.263][1][info][upstream] [source/server/lds_api.cc:55] lds: remove listener 'jenkins_test_ingress'
envoy_1  | [2020-03-17 07:23:12.266][1][info][upstream] [source/server/lds_api.cc:73] lds: add/update listener 'ads_egress'
envoy_1  | [2020-03-17 07:23:12.269][1][warning][config] [source/common/config/grpc_subscription_impl.cc:87] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) kinesis_egress: route: unknown cluster 'kinesis_cluster'
envoy_1  | 

```

**Comment by htuch** [comment]
@winguse are you using ADS? Do you have a trace of what the sequencing looks like?

**Comment by winguse** [comment]
@htuch yes, we are using ADS. and the sequence are the same as the above. But this time, we cannot reproduce this in a stable way, it happen at times. I have a test run against all ~150 node configs, it will random fail ~3 cases.

**Comment by htuch** [comment]
@winguse are you using logical DNS or EDS?

**Comment by winguse** [comment]
@htuch we are using logical DNS

**Comment by winguse** [comment]
that's a valid point, as you can see the log above, it's aws kinesis. it sounds like it's because DNS lookup taking longer time?

**Comment by htuch** [comment]
@winguse yeah, I think this is the same as https://github.com/envoyproxy/envoy/issues/5174. Did you consider enhancing your ADS to do (2) in https://github.com/envoyproxy/envoy/issues/5174#issuecomment-453689096? This should now be an option since https://github.com/envoyproxy/envoy/pull/5887 has merged.

**Comment by winguse** [comment]
@htuch we did send LDS/RDS update only after envoy acked the CDS.

**Comment by htuch** [comment]
@winguse that seems surprising, @ivitjuk is there a reason that https://github.com/envoyproxy/envoy/pull/5887 would not be the solution given what @winguse describes?

**Comment by ivitjuk** [comment]
@htuch as far as I remember, it worker well when I implemented it. I do see initial implementation was patched in de18671ab2f89ddb52adaa82e50cb592b4f69298 

@winguse from the logs in the issue description, I see you added the ranking_egress listener before datascience cluster finished warming. Are you sure you waited for the ACK in your ADS implementation? Also, I don't see the 'warming cluster ... complete' log lines in your logs.

**Comment by winguse** [comment]
@ivitjuk we did wait for Envoy to ack the CDS delivery before sending LDS updates. btw, I don's see warming cluster log,  is that require debug log?
